lib/tpm2_hmac_auth.c

```diff
@@ -0,0 +1,208 @@
+// and/or other materials provided with the distribution.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+// THE POSSIBILITY OF SUCH DAMAGE.
+//**********************************************************************;
+#ifndef HMAC_AUTH_H
+#define HMAC_AUTH_H
+
+#include <string.h>
+
+#include <tss2/tss2_sys.h>
+
+#include <openssl/bio.h>
+#include <openssl/buffer.h>
+#include <openssl/evp.h>
+#include <openssl/sha.h>
+
+#include "log.h"
+#include "tpm2_error.h"
+#include "tpm2_nv_util.h"
+#include "tpm2_openssl.h"
+#include "tpm2_session.h"
+
+uint8_t *tpm2_get_cp_hash(TSS2_SYS_CONTEXT *sapi_context, TPM2B_NAME entity_1_name,
+    TPM2B_NAME entity_2_name) {
+
+    SHA256_CTX sha256;
+    int is_success = SHA256_Init(&sha256);
+    if(!is_success) {
+        LOG_ERR("SHA256_Init failed when calculating HMAC");
+        return NULL;
+    }
+
+    uint8_t cmd_code[4];
+    TSS2_RC rval = Tss2_Sys_GetCommandCode(sapi_context, cmd_code);
+    if(rval != TPM2_RC_SUCCESS) {
+        LOG_PERR(Tss2_Sys_GetCommandCode, rval);
+        return NULL;
+    }
+
+    is_success = SHA256_Update(&sha256, cmd_code, sizeof(cmd_code));
+    if (!is_success) {
+        LOG_ERR ("SHA256_Update failed when calculating HMAC");
+        return NULL;
+    }
+
+    is_success = SHA256_Update(&sha256, entity_1_name.name, entity_1_name.size);
+    if (!is_success) {
+        LOG_ERR ("SHA256_Update failed when calculating HMAC");
+        return NULL;
+    }
+    is_success = SHA256_Update(&sha256, entity_2_name.name, entity_2_name.size);
+    if (!is_success) {
+        LOG_ERR ("SHA256_Update failed when calculating HMAC");
+        return NULL;
+    }
+
+    size_t command_params_size;
+    const uint8_t *command_params;
+    rval = Tss2_Sys_GetCpBuffer(sapi_context, &command_params_size, &command_params);
+    if (rval != TPM2_RC_SUCCESS) {
+        LOG_PERR(Tss2_Sys_GetCpBuffer, rval);
+        return NULL;
+    }
+
+    is_success = SHA256_Update(&sha256, command_params, command_params_size);
+    if (!is_success) {
+        LOG_ERR ("SHA256_Update failed when calculating HMAC");
+        return NULL;
+    }
+
+    uint8_t *cp_hash = malloc(SHA256_DIGEST_LENGTH);
+    if (!cp_hash) {
+        LOG_ERR ("OOM");
+        return NULL;
+    }
```

Comment:
you have newlines between your code blocks except for here, add it.

Reply:
fixed